test/lifecycle_test.dart

```diff
@@ -371,11 +476,28 @@ void main() {
 
         react_dom.render(LifecycleTest(newProps), mountNode);
 
-        expect(component.lifecycleCalls, equals([
-          matchCall('componentWillReceiveProps', args: [newPropsWithDefaults],           props: initialPropsWithDefaults, state: initialState),
-          matchCall('shouldComponentUpdate',     args: [newPropsWithDefaults, newState], props: initialPropsWithDefaults, state: initialState),
-        ]));
+        List calls = [
+          matchCall('componentWillReceiveProps',            args: [newPropsWithDefaults],                            props: initialPropsWithDefaults, state: initialState),
+          matchCall('componentWillReceivePropsWithContext', args: [newPropsWithDefaults, expectedContext],           props: initialPropsWithDefaults, state: initialState),
+          matchCall('shouldComponentUpdateWithContext',     args: [newPropsWithDefaults, newState, expectedContext], props: initialPropsWithDefaults, state: initialState),
+        ];
+
+        if (shouldComponentUpdateWithContext == null) {
+          calls.add(
+            matchCall('shouldComponentUpdate', args: [newPropsWithDefaults, newState], props: initialPropsWithDefaults, state: initialState),
+          );
+        }
+
+        expect(component.lifecycleCalls, equals(calls));
       });
+    }
+
+    group('when shouldComponentUpdate returns false:', () {
+      testShouldUpdates(shouldComponentUpdateWithContext: null, shouldComponentUpdate: false);
+    });
+
+    group('when shouldComponentUpdateWithContext returns false:', () {
+      testShouldUpdates(shouldComponentUpdateWithContext: false, shouldComponentUpdate: false);
     });
```

Comment:
Should we add one more test case that `shouldComponentUpdate` is not called when `shouldComponentUpdateWithContext` returns `true`? Or is that overkill?

Reply:
That's pretty much what this is testing:
```
        List calls = [
          matchCall('componentWillReceiveProps', args: [newPropsWithDefaults],                props: initialPropsWithDefaults),
          matchCall('componentWillReceivePropsWithContext', args: [newPropsWithDefaults, expectedContext],                props: initialPropsWithDefaults),
          matchCall('shouldComponentUpdateWithContext',     args: [newPropsWithDefaults, expectedState, expectedContext], props: initialPropsWithDefaults),
        ];

        if (shouldComponentUpdateWithContext == null) {
          calls.add(
            matchCall('shouldComponentUpdate', args: [newPropsWithDefaults, expectedState], props: initialPropsWithDefaults),
          );
        }
```

When `shouldComponentUpdateWithContext` is `null` we validate the method is called. If it was called when `shouldComponentUpdateWithContext` had a value, this test would fail.

Reply:
Right, I just meant testing `true` as a value in addition to `false`.